### PR TITLE
549 backend error messages

### LINKED ISSUE #484: Add date validations
For dates that are in the future change the error message to say:
- [Field name] cannot be in the future. 
- Make sure dates are in MM/DD/YYYY format
-- for example: 030/02/2020 should throw an error
- Other date validations:
- Funding period can't begin before an enrollment

### LINKED ISSUE #549: use backend validation error  messages
- Validation messages displayed in the application should come from the messages in the api responses
- error status should be displayed for hidden forms (i.e. enrollment, funding, family income det) when they are collapsed
  - right now, we only show the `!` if the field is missing. need to update this check to show the `!` if the field has a validation error.

### PR DESCRIPTION
**Pull request by mmazanec22**
Closes #549 closes #484 

Some of our funding validations weren't working because I misunderstood the nested validations when I wrote them.  I changed them to follow the same pattern we're using for child/family.

### CODE REVIEW

File: client/src/containers/EditRecord/EditRecord.tsx [hunk 1]
```diff
@@ -95,6 +95,7 @@ const EditRecord: React.FC = () => {
     afterSaveSuccess: () => setTriggerRefetchCounter((r) => r + 1),
     setAlerts,
   };
+  console.log(child);
```
Comment (belucid) [inline client/src/containers/EditRecord/EditRecord.tsx:98]: @mmazanec22 did you mean to keep this in?

Comment (mmazanec22) [reply client/src/containers/EditRecord/EditRecord.tsx:98]: Nope!

File: client/src/utils/formatters/sentenceCase.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
+export function sentenceCase(input: string) {
+  return input.charAt(0).toUpperCase() + input.toLowerCase().slice(1);
```
Comment (belucid) [inline client/src/utils/formatters/sentenceCase.ts:2]: Do we want to lower case the rest of the characters? Won't this affect intentional casing in the rest of the sentence for names, proper nouns, acronyms, etc.?

Comment (mmazanec22) [reply client/src/utils/formatters/sentenceCase.ts:2]: AFAIK there are no names or proper nouns in any of our labels

Comment (mmazanec22) [reply client/src/utils/formatters/sentenceCase.ts:2]: Ugh but there is an abbreviation-- will handle that edge case in the error message func

**Comment by belucid** [comment]
@mmazanec22 given that there aren't automated test changes to cover these error message changes, and also I know you went through some manual QA on this (we discussed it on the standup), it's good practice to include the manual tests steps you performed in the PR description. This helps the reviewer a lot, for a lot of reasons.

1) The reviewers knows manual QA has been done (a PR submitter is quite unlikely to document test steps they didn't perform)
2) It's easier to review the QA thoughtfully when it's written down... "does this cover every case... anything missing"
3) Reviewers should really be repeating any manual QA, and it's much easier and saves a lot of time if there are steps they can just follow. Repeating the manual QA helps find more issues (everyone's env. is just a little different, they perform steps with minor variations), and the going through the QA steps themselves almost always helps the reviewer with #2 above.

Ideally we'd also be using a template for PR descriptions, but that's something we can discuss as a team on the next retro.

File: client/src/containers/BatchEdit/EnrollmentFunding/__snapshots__/Form.test.tsx.snap [hunk 3]
```diff
@@ -86,7 +86,7 @@ exports[`BatchEdit EnrollmentFundingForm matches snapshot 1`] = `
         id="status-firstReportingPeriod"
         role="alert"
       >
-         First reporting period is required for OEC reporting.
+         Field is required for OEC reporting.
```
Comment (jlhogan) [inline client/src/containers/BatchEdit/EnrollmentFunding/__snapshots__/Form.test.tsx.snap:89]: does this change mean we lose the ability to drill down and get display name from html label?

File: client/src/utils/getValidationStatus.ts [hunk 4]
```diff
@@ -1,8 +1,31 @@
...
+  EnrichedValidationError,
+  ObjectWithValidationErrors,
+} from '../shared/models/ObjectWithValidationErrors';
+import { sentenceCase } from './formatters/sentenceCase';
+
+const getIsRequiredMessage = (fieldName: string) =>
+  `${fieldName} is required for OEC reporting.`;
+
+function getValidationErrorMessage(validationError: EnrichedValidationError) {
+  const { constraints } = validationError || {};
```
Comment (jlhogan) [inline client/src/utils/getValidationStatus.ts:14]: can you throw in some comments here. What is the constraint, and why does its absence mean we use the is required message?

File: client/src/utils/getValidationStatus.ts [hunk 5]
```diff
@@ -16,6 +39,7 @@ export function getValidationStatusForFieldInFieldset<
   path: string,
   fieldProps: any & { label: string }
 ): FormStatusProps | undefined {
+  // TODO: do we actually want to do this?
```
Comment (jlhogan) [inline client/src/utils/getValidationStatus.ts:42]: why not? what's up ?

File: client/src/utils/getValidationStatus.ts [hunk 6]
```diff
@@ -111,7 +124,11 @@ export function getValidationStatusForFields<
     fields.includes(v.property)
   );
   if (!validationErrors.length) return;
-  const message = 'This information is required for OEC reporting.';
+  const message = validationErrors.reduce((prev, current) => {
```
Comment (jlhogan) [inline client/src/utils/getValidationStatus.ts:127]: this is chaining multiple messages for multiple fields? When do we want to do this? what does the end result look like? String of sentences?

File: src/entity/Child.ts [hunk 7]
```diff
@@ -46,14 +46,14 @@ export class Child implements ChildInterface {
...
-  @IsNotEmpty({ message: 'First name is required' })
+  @IsNotEmpty()
   firstName?: string;
 
   @Column({ nullable: true })
   middleName?: string;
 
   @Column({ nullable: true })
-  @IsNotEmpty({ message: 'Last name is required' })
+  @IsNotEmpty()
```
Comment (jlhogan) [inline src/entity/Child.ts:56]: Ah I see so we're going with: if no message from backend, then assume it is a 'missing but required' thing and that's why no constraint in my comment above means use that error message :) not sure where, but should capture that somewhere. Maybe all the comments can livein the FE utils?

File: src/entity/Child.ts [hunk 8]
```diff
@@ -163,7 +163,8 @@ export class Child implements ChildInterface {
 
   @ValidateNested()
   @ValidateIf((child) => {
-    if (child.family) child.family.childIsFoster = child.foster;
+    if (child.family)
+      child.family.childIsFoster = child.foster === UndefinableBoolean.Yes;
```
Comment (jlhogan) [inline src/entity/Child.ts:167]: ooooo

File: src/entity/Enrollment.ts [hunk 9]
```diff
@@ -63,9 +64,17 @@ export class Enrollment implements EnrollmentInterface {
 
   @Column({ type: 'date', nullable: true, transformer: momentTransformer })
   @IsNotEmpty()
+  @MomentComparison({
+    compareFunc: (entry: Moment) => entry.isBefore(moment()),
+    message: 'Entry cannot be in the future.',
```
Comment (jlhogan) [inline src/entity/Enrollment.ts:69]: Did this come from some specific AC somewhere, or just something that seems logical?

File: src/entity/Enrollment.ts [hunk 10]
```diff
@@ -80,9 +89,22 @@ export class Enrollment implements EnrollmentInterface {
...
   @ValidateNested({ each: true })
-  @FundingDoesNotOverlap()
   @FundingAgeGroupMatchesEnrollment()
   @IsNotEmpty()
+  @ValidateIf((enrollment) => {
+    // This is used by each funding and then deleted
+    // See similar pattern between child and family
+    enrollment.fundings.forEach((funding) => {
+      // Fundings must be undefined or we have circular json
+      funding.enrollment = { ...enrollment, fundings: undefined };
```
Comment (jlhogan) [inline src/entity/Enrollment.ts:99]: does funding specifically need enrollment for some validation? or does it just need the other fundings from the enrollment ?

File: src/entity/decorators/Funding/fundingBeginsAfterEnrollmentEntry.ts [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+    registerDecorator({
+      name: 'fundingStart',
+      target: object.constructor,
+      propertyName,
+      options: { message, ...validationOptions },
+      constraints: [],
+      validator: {
+        validate(reportingPeriod: ReportingPeriod, { object: funding }) {
+          const { entry } = (funding as Funding)?.enrollment;
+          // TODO: should this be based on period start or period end???
```
Comment (jlhogan) [inline src/entity/decorators/Funding/fundingBeginsAfterEnrollmentEntry.ts:19]: the end of the first reporting period must be after the enrollment start

File: src/utils/validateObject.ts [hunk 12]
```diff
@@ -1,15 +1,22 @@
...
-  const validationErrors = await validate(object, {
+  let validationErrors = await validate(object, {
     validationError: { value: false, target: false },
   });
-  return distributeValidationErrorsToSubObjects(object, validationErrors);
+  const metadata = getAllColumnMetadata();
+  return distributeValidationErrorsToSubObjects(
+    object,
+    validationErrors,
+    metadata
```
Comment (jlhogan) [inline src/utils/validateObject.ts:18]: So we're adding the metadata just to get the formatted name? Can we no longer get that from the react node in the UI?

File: src/entity/Enrollment.ts (hunk 9 above)
Comment (mmazanec22) [reply src/entity/Enrollment.ts:69]: AC is first bullet here: https://github.com/ctoec/data-collection/issues/484

File: client/src/utils/getValidationStatus.ts (hunk 5 above)
Comment (mmazanec22) [reply client/src/utils/getValidationStatus.ts:42]: When we use this func we're not showing granular error messages for each field in the fieldset.  That makes sense for something like a date (maybe), but could be annoyingly vague in cases where a particular value is a blocker

File: client/src/utils/getValidationStatus.ts (hunk 6 above)
Comment (mmazanec22) [reply client/src/utils/getValidationStatus.ts:127]: Yep, string of sentences

File: src/entity/Enrollment.ts (hunk 10 above)
Comment (mmazanec22) [reply src/entity/Enrollment.ts:99]: Yep, for `FundingBeginsAfterEnrollmentEntry`

File: src/utils/validateObject.ts (hunk 12 above)
Comment (mmazanec22) [reply src/utils/validateObject.ts:18]: We can.  I made this change bc it seemed more stable to treat metadata as the source of truth so that there would be consistency in what the user sees across the app, between data requirements, the template, and our forms.  Was wondering even if we should consider using metadata as the default for field labels, and eventually for help text if we add that in.  As it stands there are a bunch of places we need to update (front end, metdata, error messages created in the backend) when terminology changes.  Can revert this if we're worried about label/error message consistency though!

Comment (jlhogan) [reply src/utils/validateObject.ts:18]: no that concerned. I think that's a smart idea -- that this should be the source of truth for everything (obviously, not to implement in full right now, but to set up for for thefuture). Given that, I  wonder if shoving _ALL_ the metadata in each validation error is necessary (theres a lot of text in there). Might be worth just popping the formatted and property names on here as opposed to the whole object?

Comment (mmazanec22) [reply src/utils/validateObject.ts:18]: That's fair!  Given the other comments about capitalization I wonder if we should change the metadata to default to not capitalized, then capitalize it where we use it if necessary (i.e. title casing in the template if we want that and sentence casing in the front end).  Would that break anything if I snuck it into this PR?

Comment (jlhogan) [reply src/utils/validateObject.ts:18]: If the names get title-cased in the template, then that'll need to be accounted for here: https://github.com/ctoec/data-collection/blob/base/src/controllers/enrollmentReports/parse.ts#L289 (prob just make the comparison of headers case-insensitive by lower-casing the ones from the template?)

File: client/src/utils/getValidationStatus.ts (hunk 5 above)
Comment (jlhogan) [reply client/src/utils/getValidationStatus.ts:47]: well i think when we want to display the full error on the field, just use getValidationStatusForField, right? like this is specifically for when we want the message-less validation status (and maybe should be renamed as such?)

File: src/template/constants.ts [hunk 13]
```diff
@@ -4,7 +4,7 @@ import moment from 'moment';
 // TODO: add pre-merge hook that updates these values (probably put them in a diff file)
 // automatically when this file gets updated
 export const TEMPLATE_VERSION = 1;
-export const TEMPLATE_LAST_UPDATED = moment.utc('11-24-2020', ['MM-DD-YYYY']);
+export const TEMPLATE_LAST_UPDATED = moment.utc('12-16-2020', ['MM-DD-YYYY']);
```
Comment (jlhogan) [inline src/template/constants.ts:7]: :raised_hands:

**Review by jlhogan — APPROVED** [review]

**Review by bamader — APPROVED** [review]
Wow that's comprehensive. Looks great!
